crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -253,6 +253,74 @@ impl<'w> EntityMut<'w> {
         self
     }
 
+    /// Inserts a component with the given `value`. Will replace the value if it already existed.
+    ///
+    /// **You should prefer to use the typed API [`EntityMut::insert`] where possible and only
+    /// use this in cases where the actual types are not known at compile time.**
+    ///
+    /// # Safety
+    /// The value referenced by `value` must be valid for the given [`ComponentId`] of this world
+    pub unsafe fn insert_by_id(
+        &mut self,
+        component_id: ComponentId,
+        value: OwningPtr<'_>,
+    ) -> &mut Self {
+        self.insert_bundle_by_ids(vec![component_id], vec![value])
+    }
+
+    /// Inserts a bundle of components into the entity. Will replace the values if they already existed.
+    ///
+    /// **You should prefer to use the typed API [`EntityMut::insert_bundle`] where possible and only
+    /// use this in cases where the actual types are not known at compile time.**
+    ///
+    /// # Safety
+    /// - each value of `components` must be valid for the [`ComponentId`] at the matching position in `component_ids` in this world
+    pub unsafe fn insert_bundle_by_ids(
+        &mut self,
+        mut component_ids: Vec<ComponentId>,
+        components: Vec<OwningPtr<'_>>,
+    ) -> &mut Self {
```

Comment:
We can add a generic param for `component_ids` and `components` and `T: IntoIterator<Item = Componentid>` and `for<'a> &'a T: IntoIterator<Item = &'a ComponentId>` (replace  ComponentId with OwningPtr for the other generic) as a temporary solution to allow calling this with arrays. This won't let you call it with `Box<[OwningPtr<'_>]>` though which is unfortunate since `&Box<[OwningPtr<'_>]>: IntoIterator` isnt a thing...

We might want to think about adding a `OwningPtr` to `bevy_ptr` that has a type parameter? Then we can have the trait impls on `&TypedOwningPtr<'_, [OwningPtr<'_>]>` and pass that in to this function :thinking: 

At the very least just making this generic will remove allocs for the case of `insert_by_id` which seems like a win even if its not an ideal situation :upside_down_face: I hope someone else can figure out a good thing to accept here?

---

alt idea: instead of relying on trait impls we could take as args:
- `impl FnOnce(OwningPtrThingy) -> impl Iterator<Item = OwningPtr<'_>>`
- `impl Fnonce(&OwningPtrThingy) -> impl Iterator<Item = &OwningPtr<'_>>`

`foo.insert_bundle_by_ids(..., [ptr1], |ptrs| ptrs.into_iter(), |ptrs| ptrs.iter())`